receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go

```diff
@@ -0,0 +1,435 @@
+		return true
+	})
+	return fmt.Sprintf("%s,%s", name, strings.Join(labelValues, ","))
+}
+
+// Remove timeseries that have aged out.
+func (tsm *timeseriesMapPdata) gc() {
+	tsm.Lock()
+	defer tsm.Unlock()
+	// this shouldn't happen under the current gc() strategy
+	if !tsm.mark {
+		return
+	}
+	for ts, tsi := range tsm.tsiMap {
+		if !tsi.mark {
+			delete(tsm.tsiMap, ts)
+		} else {
+			tsi.mark = false
+		}
+	}
+	tsm.mark = false
+}
+
+func newTimeseriesMapPdata() *timeseriesMapPdata {
+	return &timeseriesMapPdata{mark: true, tsiMap: map[string]*timeseriesinfoPdata{}}
+}
+
+// JobsMapPdata maps from a job instance to a map of timeseriesPdata instances for the job.
+type JobsMapPdata struct {
+	sync.RWMutex
+	gcInterval time.Duration
+	lastGC     time.Time
+	jobsMap    map[string]*timeseriesMapPdata
+}
+
+// NewJobsMap creates a new (empty) JobsMapPdata.
+func NewJobsMapPdata(gcInterval time.Duration) *JobsMapPdata {
+	return &JobsMapPdata{gcInterval: gcInterval, lastGC: time.Now(), jobsMap: make(map[string]*timeseriesMapPdata)}
+}
+
+// Remove jobs and timeseries that have aged out.
+func (jm *JobsMapPdata) gc() {
+	jm.Lock()
+	defer jm.Unlock()
+	// once the structure is locked, confirm that gc() is still necessary
+	if time.Since(jm.lastGC) > jm.gcInterval {
+		for sig, tsm := range jm.jobsMap {
+			tsm.RLock()
+			tsmNotMarked := !tsm.mark
+			tsm.RUnlock()
+			if tsmNotMarked {
+				delete(jm.jobsMap, sig)
+			} else {
+				tsm.gc()
+			}
+		}
+		jm.lastGC = time.Now()
+	}
+}
+
+func (jm *JobsMapPdata) maybeGC() {
+	// speculatively check if gc() is necessary, recheck once the structure is locked
+	jm.RLock()
+	defer jm.RUnlock()
+	if time.Since(jm.lastGC) > jm.gcInterval {
+		go jm.gc()
+	}
+}
+
+func (jm *JobsMapPdata) get(job, instance string) *timeseriesMapPdata {
+	sig := job + ":" + instance
+	jm.RLock()
+	tsm, ok := jm.jobsMap[sig]
+	jm.RUnlock()
+	defer jm.maybeGC()
+	if ok {
+		return tsm
+	}
+	jm.Lock()
+	defer jm.Unlock()
+	tsm2, ok2 := jm.jobsMap[sig]
+	if ok2 {
+		return tsm2
+	}
+	tsm2 = newTimeseriesMapPdata()
+	jm.jobsMap[sig] = tsm2
+	return tsm2
+}
+
+// MetricsAdjusterPdata takes a map from a metric instance to the initial point in the metrics instance
+// and provides AdjustMetrics, which takes a sequence of metrics and adjust their start times based on
+// the initial points.
+type MetricsAdjusterPdata struct {
+	tsm    *timeseriesMapPdata
+	logger *zap.Logger
+}
+
+// NewMetricsAdjuster is a constructor for MetricsAdjuster.
+func NewMetricsAdjusterPdata(tsm *timeseriesMapPdata, logger *zap.Logger) *MetricsAdjusterPdata {
+	return &MetricsAdjusterPdata{
+		tsm:    tsm,
+		logger: logger,
+	}
+}
+
+// AdjustMetrics takes a sequence of metrics and adjust their start times based on the initial and
+// previous points in the timeseriesMap.
+// Returns the total number of timeseries that had reset start times.
+func (ma *MetricsAdjusterPdata) AdjustMetrics(metricL *pdata.MetricSlice) int {
+	resets := 0
+	ma.tsm.Lock()
```

Comment:
It's somewhat mysterious we're locking on `tsm` here but not accessing it (presumably we are accessing through `adjustMetric`). Maybe a comment to ensure it was intended